### ISSUE TITLE
"Cannot release a lock that's no longer owned" - redis.exceptions.LockNotOwnedError

### ISSUE THREAD

**Issue by matt-schwartz** [PostHog/plugin-server#237]
## Bug description

I started using Posthog in my production environment 4 days ago (loving it, by the way).  I've seen the following Redis lock error in my logs a couple of times: 

```
x.x.x.x - - [10/Mar/2021:11:15:05 +0000] "POST /batch/ HTTP/1.1" 200 13 "-" "posthog-python/1.2.1"
x.x.x.x - - [10/Mar/2021:11:15:07 +0000] "POST /batch/ HTTP/1.1" 200 13 "-" "posthog-python/1.2.1"
[2021-03-10 11:16:04 +0000] [13] [CRITICAL] WORKER TIMEOUT (pid:36)
[2021-03-10 11:18:17 +0000] [13] [CRITICAL] WORKER TIMEOUT (pid:35)
[2021-03-10 11:19:47 +0000] [79] [INFO] Booting worker with pid: 79
[2021-03-10 11:19:47 +0000] [80] [INFO] Booting worker with pid: 80
[MAIN] 🔴 Redlock cannot extend lock:
LockError: Cannot extend lock on resource "plugin-server:locks:schedule" because the lock has already expired.
at Redlock.extend (/code/plugins/node_modules/redlock/redlock.js:299:25)
at Lock.extend (/code/plugins/node_modules/redlock/redlock.js:106:22)
at Timeout.extendLock [as _onTimeout] (/code/plugins/node_modules/@posthog/plugin-server/dist/services/schedule.js:63:39)
at listOnTimeout (internal/timers.js:554:17)
at processTimers (internal/timers.js:497:7) {
attempts: 0
}
[MAIN] 🔒 Scheduler lock acquired!
[MAIN] ✅ Connected to Redis!
[2021-03-10 11:19:48,124: CRITICAL/MainProcess] beat raised exception <class 'redis.exceptions.LockNotOwnedError'>: LockNotOwnedError("Cannot release a lock that's no longer owned")
Traceback (most recent call last):
File "/usr/local/lib/python3.8/site-packages/celery/beat.py", line 631, in start
interval = self.scheduler.tick()
File "/usr/local/lib/python3.8/site-packages/redbeat/schedulers.py", line 466, in tick
self.lock.extend(int(self.lock_timeout))
File "/usr/local/lib/python3.8/site-packages/redis/lock.py", line 246, in extend
return self.do_extend(additional_time)
File "/usr/local/lib/python3.8/site-packages/redis/lock.py", line 253, in do_extend
raise LockNotOwnedError("Cannot extend a lock that's"
redis.exceptions.LockNotOwnedError: Cannot extend a lock that's no longer owned
During handling of the above exception, another exception occurred:
Traceback (most recent call last):
File "/usr/local/lib/python3.8/site-packages/celery/apps/beat.py", line 109, in start_scheduler
service.start()
File "/usr/local/lib/python3.8/site-packages/celery/beat.py", line 641, in start
self.sync()
File "/usr/local/lib/python3.8/site-packages/celery/beat.py", line 644, in sync
self.scheduler.close()
File "/usr/local/lib/python3.8/site-packages/redbeat/schedulers.py", line 482, in close
self.lock.release()
File "/usr/local/lib/python3.8/site-packages/redis/lock.py", line 226, in release
self.do_release(expected_token)
File "/usr/local/lib/python3.8/site-packages/redis/lock.py", line 232, in do_release
raise LockNotOwnedError("Cannot release a lock"
redis.exceptions.LockNotOwnedError: Cannot release a lock that's no longer owned
```

## Expected behavior

No exception raised.

## How to reproduce

1. Run docker image `redis:5.0-alpine`
2. Run docker image `posthog:latest` on the same machine with `REDIS_URL=redis://redis:6379/2`
3. Let the system run and pick up events for a while.

## Environment

- Amazon EC2 t3.small instance (2 CPU, 2GB RAM)
- posthog:latest docker image (v1.22.0)

## Additional context

Note I'm connecting to redis database "2".  I doubt that would make a difference, but just in case...

**Comment by yakkomajuri** [comment]
It's definitely not nice that we throw this error, but it shouldn't prevent anything from running - is everything working ok for you?

**Comment by matt-schwartz** [comment]
@yakkomajuri Everything seems to be running fine.  It appears all data is still being collected and aggregated.

**Comment by mariusandra** [comment]
Reading the log, what I find interesting is that both a) the python celery process and b) the nodejs plugin server... both got the same error after one another, even though they're two different programs. They're both using a distributed locking system called [redlock](https://redis.io/topics/distlock) to make sure there's only one server starting scheduled tasks at any one time.

This scheduler lock is given for just a minute, and to keep it longer, you must continuously extend it. The error here shows that the lock had expired when there was an attempt to extend it. This might just mean that the server was clogged at the time. For example some other service was taking all the resources, or a lot of swapping was taking place, or ...? Not sure. 

Your instance size should be fine, if a bit on the lower end. Do you also have a postgres docker container in there?

**Comment by matt-schwartz** [comment]
@mariusandra I'm running 3 docker containers on this server: Posthog, redis, and my primary web application (a Django app).  Swap is never hit and memory consumption remains consistent around 80%.  CPU stays below 5% except for one or two short spikes to ~50% some days.

PH and my web app both connect to Postgres on RDS.

I'll continue to monitor and see if the error correlates with spikes in CPU, network, memory, or any other resources.

Is it possible there's a default redis setting that would time-out the lock request?  The system is always busy logging API and other web requests, so I doubt anything would timeout, just a thought.

**Comment by matt-schwartz** [comment]
Happened again and I noticed after carefully checking CloudWatch logs that the last few times coincided with a memory and CPU spike.  When memory is about to max out redis will dump data (in a default setup).  My guess is this is related.

I've moved Posthog to it's own server and will increase memory if needed.  Closing this ticket since it's definitely not a code problem.

And thanks again for both the support and excellent open source project!
